### PR TITLE
feat: Initial support for Gateway API

### PR DESCRIPTION
**Pull request by tao12345666333**
Please answer these questions before submitting a pull request

- Why submit this pull request?

    - [x] New feature provided

- Related issues

    #644 
___

### New feature or improvement

- Describe the details and related test reports.

    In this version, the initial Gateway API support has been added, and a switch has been added to control whether to enable this feature. So it can be safely merged.

     Changes in this PR:

         * Initial support for Gateway API
         * Update golang to v1.16

    Note: This feature is still under development, and the route has not yet been published to the Apache APISIX Gateway

### CODE REVIEW

File: cmd/ingress/ingress.go [hunk 1]
```diff
@@ -158,6 +158,7 @@ For example, no available LB exists in the bare metal environment.`)
 	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.IngressVersion, "ingress-version", config.IngressNetworkingV1, "the supported ingress api group version, can be \"networking/v1beta1\", \"networking/v1\" (for Kubernetes version v1.19.0 or higher) and \"extensions/v1beta1\"")
 	cmd.PersistentFlags().StringVar(&cfg.Kubernetes.ApisixRouteVersion, "apisix-route-version", config.ApisixRouteV2beta3, "the supported apisixroute api group version, can be \"apisix.apache.org/v2beta1\" or \"apisix.apache.org/v2beta2\" or \"apisix.apache.org/v2beta3\"")
 	cmd.PersistentFlags().BoolVar(&cfg.Kubernetes.WatchEndpointSlices, "watch-endpointslices", false, "whether to watch endpointslices rather than endpoints")
+	cmd.PersistentFlags().BoolVar(&cfg.Kubernetes.EnableGateway, "enable-gateway", false, "whether to enable support for Gateway API")
```
Comment (tokers) [inline cmd/ingress/ingress.go:161]:
```suggestion
	cmd.PersistentFlags().BoolVar(&cfg.Kubernetes.EnableGateway, "enable-gateway-api", false, "whether to enable support for Gateway API")
```

File: pkg/ingress/gateway.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		return
+	}
+	if !c.controller.namespaceWatching(key) {
+		return
+	}
+	log.Debugw("gateway add event arrived",
+		zap.Any("object", obj),
+	)
+
+	c.workqueue.AddRateLimited(&types.Event{
```
Comment (gxthrj) [inline pkg/ingress/gateway.go:150]:
```suggestion
	c.workqueue.Add(&types.Event{
```

**Comment by tao12345666333** [comment]
The CI failed.

```
Step 9/17 : COPY . .
 ---> 7d2b8b4e52e3
Step 10/17 : RUN make build
 ---> Running in 4af72474093a
go build \
	-o apisix-ingress-controller \
	-ldflags "-X="github.com/apache/apisix-ingress-controller/pkg/version._buildVersion"=1.3.0 -X="github.com/apache/apisix-ingress-controller/pkg/version._buildGitRevision"="no-git-module" -X="github.com/apache/apisix-ingress-controller/pkg/version._buildOS"=linux/amd64" \
	main.go
Error: /go/pkg/mod/golang.org/x/net@v0.0.0-20211208012354-db4efeb81f4b/idna/idna10.0.0.go:26:2: missing go.sum entry for module providing package golang.org/x/text/secure/bidirule; to add:
	go mod download golang.org/x/text
Error: /go/pkg/mod/golang.org/x/net@v0.0.0-20211208012354-db4efeb81f4b/idna/idna10.0.0.go:27:2: missing go.sum entry for module providing package golang.org/x/text/unicode/bidi; to add:
	go mod download golang.org/x/text
Error: /go/pkg/mod/golang.org/x/net@v0.0.0-20211208012354-db4efeb81f4b/idna/idna10.0.0.go:28:2: missing go.sum entry for module providing package golang.org/x/text/unicode/norm; to add:
	go mod download golang.org/x/text
Error: make: *** [Makefile:46: build] Error 1
The command '/bin/sh -c make build' returned a non-zero code: 2
Error: make: *** [Makefile:100: push-images-to-kind] Error 2

Error: Process completed with exit code 2.
```

This error may be related to the environment in the GitHub action.

File: .github/workflows/lint-checker.yml [hunk 3]
```diff
@@ -64,10 +64,10 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
-      - name: setup go
-        uses: actions/setup-go@v1
+      - name: Setup Go Env
+        uses: actions/setup-go@v2
         with:
-          go-version: '1.13'
+          go-version: "1.16.8"
```
Comment (tokers) [inline .github/workflows/lint-checker.yml:70]: I think we should use the latest patch version of go 1.16 version, as there are some security fixes after go 1.16.8.

> go1.16.9 (released 2021-10-07) includes a security fix to the linker and misc/wasm directory, as well as bug fixes to the runtime and to the text/template package. See the Go 1.16.9 milestone on our issue tracker for details.

> go1.16.10 (released 2021-11-04) includes security fixes to the archive/zip and debug/macho packages, as well as bug fixes to the compiler, linker, runtime, the misc/wasm directory, and to the net/http package. See the Go 1.16.10 milestone on our issue tracker for details.

> go1.16.11 (released 2021-12-02) includes fixes to the compiler, runtime, and the net/http, net/http/httptest, and time packages. See the Go 1.16.11 milestone on our issue tracker for details.

File: go.mod [hunk 4]
```diff
@@ -1,31 +1,25 @@
...
 	golang.org/x/sys v0.0.0-20210819072135-bce67f096156 // indirect
-	golang.org/x/tools v0.1.5 // indirect
 	gopkg.in/yaml.v2 v2.4.0
-	k8s.io/api v0.21.1
-	k8s.io/apimachinery v0.21.1
-	k8s.io/client-go v0.21.1
-	k8s.io/code-generator v0.21.1
+	k8s.io/api v0.22.4
+	k8s.io/apimachinery v0.22.4
+	k8s.io/client-go v0.22.4
```
Comment (gxthrj) [inline go.mod:22]: Need to re-run `make codegen` to see if there are any differents.

File: pkg/ingress/gateway.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		if quit {
+			return
+		}
+		err := c.sync(ctx, obj.(*types.Event))
+		c.workqueue.Done(obj)
+		c.handleSyncErr(obj, err)
+	}
+}
+
+func (c *gatewayController) sync(ctx context.Context, ev *types.Event) error {
```
Comment (gxthrj) [inline pkg/ingress/gateway.go:82]: I can not see the logic for syncing to APISIX in `sync` method.

File: go.mod (hunk 4 above)
Comment (tao12345666333) [reply go.mod:22]: There is no change. `make codegen` has been executed in this PR, and it has passed the verification in the test case.

File: pkg/ingress/gateway.go (hunk 5 above)
Comment (tao12345666333) [reply pkg/ingress/gateway.go:82]: As I wrote in the description of the PR.

> Note: This feature is still under development, and the route has not yet been published to the Apache APISIX Gateway

Only initial support in this PR. Mainly to implement the synchronization of the Gateway resource status.
Once all features are submitted through a PR, then this PR will be huge. @gxthrj

**Comment by tao12345666333** [comment]
ping @gxthrj @tokers for review, thanks!

File: pkg/ingress/gateway.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+		AddFunc:    ctl.onAdd,
+		UpdateFunc: ctl.onUpdate,
+		DeleteFunc: ctl.OnDelete,
+	})
+	return ctl
+}
+
+func (c *gatewayController) run(ctx context.Context) {
+	log.Info("gateway controller started")
+	defer log.Infof("gateway controller exited")
```
Comment (tokers) [inline pkg/ingress/gateway.go:55]:
```suggestion
	defer log.Info("gateway controller exited")
```

File: pkg/ingress/gateway.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	defer c.workqueue.ShutDown()
+
+	log.Info("gateway controller started")
```
Comment (tokers) [inline pkg/ingress/gateway.go:58]: duplicated logging.

File: pkg/ingress/gateway.go (hunk 6 above)
Comment (tao12345666333) [reply pkg/ingress/gateway.go]: Thanks!

Comment (tao12345666333) [reply pkg/ingress/gateway.go]: Thanks!

**Review by tokers — APPROVED** [review]

**Review by gxthrj — APPROVED** [review]

**Pull request merged**